lib/dictBuilder/cover.c

```diff
@@ -929,6 +929,112 @@ void COVER_best_finish(COVER_best_t *best, size_t compressedSize,
+  memcpy(largestDictbuffer, customDictContent, dictContentSize);
+  dictContentSize = ZDICT_finalizeDictionary(
+    largestDictbuffer, dictContentSize, (const void*)customDictContent, dictContentSize,
+    samplesBuffer, samplesSizes, (unsigned)nbFinalizeSamples, params.zParams);
+
+  if (ZDICT_isError(dictContentSize)) {
+    free(largestDictbuffer);
+    free(candidateDictBuffer);
+    return COVER_dictSelectionError(dictContentSize);
+  }
+
+  totalCompressedSize = COVER_checkTotalCompressedSize(params, samplesSizes,
+                                                       (const BYTE*)samplesBuffer, offsets,
+                                                       nbCheckSamples, nbSamples,
+                                                       (BYTE *const)largestDictbuffer, dictContentSize);
+
+  if (ZSTD_isError(totalCompressedSize)) {
+    free(largestDictbuffer);
+    free(candidateDictBuffer);
+    return COVER_dictSelectionError(totalCompressedSize);
+  }
+
+  if (params.shrinkDict == 0) {
+    const COVER_dictSelection_t selection = { largestDictbuffer, dictContentSize, totalCompressedSize };
+    free(candidateDictBuffer);
+    return selection;
+  }
+
+  largestDict = dictContentSize;
+  largestCompressed = totalCompressedSize;
+  dictContentSize = ZDICT_DICTSIZE_MIN;
+
+  /* Largest dict is initially at least ZDICT_DICTSIZE_MIN */
+  while (dictContentSize < largestDict) {
+    {
+      memcpy(candidateDictBuffer, largestDictbuffer, largestDict);
+      dictContentSize = ZDICT_finalizeDictionary(
+        candidateDictBuffer, dictContentSize, (const void*)(customDictContentEnd - dictContentSize), dictContentSize,
+        samplesBuffer, samplesSizes, (unsigned)nbFinalizeSamples, params.zParams);
+
+      if (ZDICT_isError(dictContentSize)) {
+        free(largestDictbuffer);
+        free(candidateDictBuffer);
+        return COVER_dictSelectionError(dictContentSize);
+      }
+    }
+
+    totalCompressedSize = COVER_checkTotalCompressedSize(params, samplesSizes,
+                                                         (const BYTE*)samplesBuffer, offsets,
+                                                         nbCheckSamples, nbSamples,
+                                                         (BYTE *const)candidateDictBuffer, dictContentSize);
+
+    if (ZSTD_isError(totalCompressedSize)) {
+      free(largestDictbuffer);
+      free(candidateDictBuffer);
+      return COVER_dictSelectionError(totalCompressedSize);
+    }
+
+    if (totalCompressedSize <= largestCompressed * regressionTolerance) {
+      const COVER_dictSelection_t selection = { candidateDictBuffer, dictContentSize, totalCompressedSize };
+      free(largestDictbuffer);
+      return selection;
+    }
+    dictContentSize *= 2;
+  }
+  dictContentSize = largestDict;
+  totalCompressedSize = largestCompressed;
+  {
+    const COVER_dictSelection_t selection = { candidateDictBuffer, dictContentSize, totalCompressedSize };
```

Comment:
Is this a bug?
1. We're returning the `candidateDictBuffer`, but giving it the size/compressed size of `largestDict`.
2. Shouldn't we return `largestDict` if we didn't find any smaller ones that were better?